packages/smoldot-provider/src/SmoldotProvider.ts

```diff
@@ -0,0 +1,346 @@
+// Copyright 2018-2020 @paritytech/substrate-connect authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import {RpcCoder} from '@polkadot/rpc-provider/coder';
+import {
+  JsonRpcResponse,
+  ProviderInterface,
+  ProviderInterfaceCallback,
+  ProviderInterfaceEmitCb,
+  ProviderInterfaceEmitted,
+} from '@polkadot/rpc-provider/types';
+import { assert, isUndefined, logger } from '@polkadot/util';
+import EventEmitter from 'eventemitter3';
+import * as smoldot from 'smoldot';
+import Database from './Database';
+
+const l = logger('smoldot-provider');
+
+interface RpcStateAwaiting {
+  callback: ProviderInterfaceCallback;
+  method: string;
+  params: any[];
+  subscription?: SubscriptionHandler;
+}
+
+interface SubscriptionHandler {
+  callback: ProviderInterfaceCallback;
+  // type is the value of the method property in the JSON responses to this
+  // subscription
+  type: string;
+}
+
+interface StateSubscription extends SubscriptionHandler {
+    method: string;
+    params: any[];
+}
+
+const ANGLICISMS: { [index: string]: string } = {
+  chain_finalisedHead: 'chain_finalizedHead',
+  chain_subscribeFinalisedHeads: 'chain_subscribeFinalizedHeads',
+  chain_unsubscribeFinalisedHeads: 'chain_unsubscribeFinalizedHeads'
+};
+/**
+ * @name SmoldotProvider
+ *
+ * @description The SmoldotProvider allows interacting with a smoldot-based
+ * WASM light client.  I.e. without doing RPC to a remote server over HTTP
+ * or websockets
+ * 
+ * @example
+ * <BR>
+ *
+ * ```javascript
+ * import readFileSync from 'fs';
+ * import Api from '@polkadot/api/promise';
+ * import { SmoldotPrpovider } from '../';
+ *
+ * const chainSpec = readFileSync('./path/to/chainSpec.json');
+ * const provider = new SmoldotProvider(chainSpec);
+ * const api = new Api(provider);
+ * ```
+ */
+export class SmoldotProvider implements ProviderInterface {
+  #chainSpec: string;
+  #coder: RpcCoder;
+  #eventemitter: EventEmitter;
+  #isConnected = false;
+  #client: smoldot.SmoldotClient | undefined = undefined;
+  #db: Database;
+  // reference to the smoldot module so we can defer loading the wasm client
+  // until connect is called
+  #smoldot: smoldot.Smoldot;
+  readonly #handlers: Record<string, RpcStateAwaiting> = {};
+  #subscriptions: Record<string, StateSubscription> = {};
+  readonly #waitingForId: Record<string, JsonRpcResponse> = {};
+
+   /**
+   * @param {string}   chainSpec  The chainSpec for the WASM client
+   * @param {Database} db         An implementation of Database for saving the chain state
```

Comment:
Is this necessary for the Provider to work or just a node specific thing?

Reply:
It's necessary.  I've provided a `Database` implementation that just saves the JSON to disk.  When we're ready to test I'll write one that saves it to localStorage

Reply:
Databases are also something completely optional, for what it's worth.
We intend to make syncing very fast even if you don't have any existing database. Having a database might save some time, but it's clear how much. If it turns out that there's no difference between database vs no database, we could even remove databases entirely.

Reply:
If it stays it would probably just make sense to find a cross browser/ Node.js solution. 
@raoulmillais  Just out of curiosity: Have you tested in a browser environment or just in node so far?

Reply:
I haven't tried in the browser actually but I'm assuming that I will just get the same websocket security error you saw when trying the polkadot browser demo based provider you wrote.  I agree a cross platform solution would be nice and it'd be easy to do.  I introduced the interface to make typescript happy and so I could inject a dummy do-nothing test database in my unit tests.